packages/SwingSet/test/virtualObjects/test-weakcollections.js

```diff
@@ -0,0 +1,153 @@
+      },
+    },
+  };
+
+  const hostStorage = provideHostStorage();
+  const bootstrapResult = await initializeSwingset(config, [], hostStorage);
+  const c = await makeSwingsetController(hostStorage, {});
+  const nextLog = makeNextLog(c);
+
+  await c.run();
+  t.deepEqual(c.kpResolution(bootstrapResult), capargs('bootstrap done'));
+
+  async function doSimple(method) {
+    const sendArgs = capargs([], []);
+    const r = c.queueToVatExport('bootstrap', 'o+0', method, sendArgs);
+    await c.run();
+    t.is(c.kpStatus(r), 'fulfilled');
+    return c.kpResolution(r);
+  }
+
+  const preGCResult = await doSimple('runProbes');
+  t.deepEqual(preGCResult, capargs('probes done'));
+  t.deepEqual(nextLog(), [
+    'probe of sample-object returns imported item #0',
+    'probe of [object Promise] returns imported item #1',
+    'probe of [Alleged: remember-exp] returns mer',
+    'probe of [Alleged: holder-vo] returns mevo',
+    'probe of [object Promise] returns mep',
+    'probe of [Alleged: forget-exp] returns fer',
+    'probe of [Alleged: holder-vo] returns fevo',
+    'probe of [object Promise] returns fep',
+  ]);
+  await doSimple('betweenProbes');
+  gc();
+  const postGCResult = await doSimple('runProbes');
+  t.deepEqual(postGCResult, capargs('probes done'));
+  t.deepEqual(nextLog(), [
+    'probe of sample-object returns imported item #0',
+    'probe of [object Promise] returns undefined',
+    'probe of [Alleged: remember-exp] returns mer',
+    'probe of [Alleged: holder-vo] returns mevo',
+    'probe of [object Promise] returns mep',
+    'probe of [Alleged: forget-exp] returns fer',
+    'probe of [Alleged: holder-vo] returns fevo',
+    'probe of [object Promise] returns fep',
+  ]);
+});
+
+test('weakMap vref handling', async t => {
+  insistGC(t);
+  const log = [];
+  const {
+    VirtualObjectAwareWeakMap,
+    VirtualObjectAwareWeakSet,
+    valToSlot,
+    slotToVal,
+  } = makeFakeVirtualObjectManager(3, log);
+
+  function addCListEntry(slot, val) {
+    slotToVal.set(slot, val);
+    valToSlot.set(val, slot);
+  }
+
+  function removeCListEntry(slot, val) {
+    slotToVal.delete(slot);
+    valToSlot.delete(val);
+  }
+
+  const weakMap = new VirtualObjectAwareWeakMap();
+
+  function checkMap(vref, label, useVRef) {
+    const obj = {};
+    addCListEntry(vref, obj);
+    weakMap.set(obj, label);
+    t.is(weakMap.get(obj), label);
+    removeCListEntry(vref, obj);
+    if (useVRef) {
+      t.falsy(weakMap.has(obj));
+      t.is(weakMap.get(obj), undefined);
+    } else {
+      t.truthy(weakMap.has(obj));
+      t.is(weakMap.get(obj), label);
+    }
+  }
+
+  checkMap('o-1', 'imported presence', true);
+  checkMap('o+2', 'exported remotable', false);
+  checkMap('o+3/4', 'exported virtual object', true);
+  checkMap('p-5', 'imported promise', false);
+  checkMap('p+6', 'exported promise', false);
+  checkMap('d-7', 'imported device', false);
+
```

Comment:
We should do an additional test here: if `useVRef` is true, then a new+different JS `Object` (with the same vref mapping in `valToSlot`, i.e. a new incarnation of a Presence or Representative) should map to the same thing. So inside `checkMap`, after `removeCListEntry`, do something like:

```js
const obj2 = {};
addCListEntry(vref, obj2);
```

then in `if (useVRef)`:
```js
t.truthy(weakMap.has(obj2));
t.is(weakMap.get(obj2), label);
```

and in `else`:
```js
t.falsy(weakMap.has(obj2));
t.is(weakMap.get(obj2), undefined);
```

and similar things in the WeakSet test.